crypto/lhash/lhash.c

```diff
@@ -97,28 +117,30 @@ void *OPENSSL_LH_insert(OPENSSL_LHASH *lh, void *data)
     OPENSSL_LH_NODE *nn, **rn;
     void *ret;
 
-    lh->error = 0;
+    clear_error(lh);
     if ((lh->up_load <= (lh->num_items * LH_LOAD_MULT / lh->num_nodes)) && !expand(lh))
-        return NULL;        /* 'lh->error++' already done in 'expand' */
+        return NULL;        /* set_error() already done in 'expand' */
 
     rn = getrn(lh, data, &hash);
 
     if (*rn == NULL) {
         if ((nn = OPENSSL_malloc(sizeof(*nn))) == NULL) {
-            lh->error++;
+            set_error(lh);
             return NULL;
         }
         nn->data = data;
         nn->next = NULL;
         nn->hash = hash;
         *rn = nn;
         ret = NULL;
-        lh->num_insert++;
         lh->num_items++;
+        if (lh->flags & OPENSSL_LH_FLAGS_STATS)
+            lh->stats.insert++;
```

Comment:
I don't understand why num_items++ is not a problem and num_insert++ is.

Reply:
Because the documentation says that if the code isn't thread-safe, so doing inserts should only happen under a lock.

I understand the concern about this being late, but it *is* a bug and this was actually the least-disruptive way I could think of to do it.  I did not want to pay the cost of making all the stats variables, which we want to discourage, be always protected internally with a lock

Reply:
So you're saying that in this function it would never have been a problem, because it should be under a lock, but that other functions that update the stats only need a lock when they want stats?

Reply:
Yes, exactly  That's what our docs have said.

Reply:
Errrr... exactly which function is it that suddenly becomes thread safe if you don't record statistics?

Reply:
Ah, it's about write lock vs. read lock...